Macvendor

aequitas opened the pull request:
Reopened from: https://github.com/home-assistant/home-assistant/pull/4401

Now uses asynchronous http calls to macvendors.com to lookup vendor string for mac on known devices.

Commit:
Add MAC vendor lookup for device_tracker.

Commit:
Test vendor mac lookup and fix device attribute.

Commit:
Generate requirements.

Commit:
Style.

Commit:
Use hyphen instead of underscore to satisfy 'idna'.

https://github.com/kjd/idna/issues/17

aequitas committed:
Resort imports.

aequitas committed:
Refactor macvendor to use macvendors.com API instead of netaddr library.

aequitas committed:
Test vendor lookup using macvendors.com api.

aequitas committed:
Remove debugging.

aequitas committed:
Correct description.

aequitas committed:
No longer needed.

lwis commented:
Can we move isort to a separate PR?

aequitas commented:
@lwis sure will do

Landrash commented:
Doesn't this change make HA dependent on a working internet connection?

aequitas committed:
Device tracker is now an async component. Fix ddwrt tests.

aequitas commented:
Had to modify the ddwrt tests suite as it was not adapted for asyncio.

I have tested it, but no http api calls are actually being made during the test suite. Btu if you want I can add a some code to make sure calls get mocked anyways, something like:

``` python
    def run(self, result=None):
        with mock_aiohttp_client() as aioclient_mock:
            aioclient_mock.get('http://api.macvendors.com/.*', text='')

            super().run(result)
```

aequitas commented:
@Landrash it fallback gracefully, so in case of no internet connection there is no harm.

aequitas committed:
Fix linting.

aequitas commented:
isort stuff is now also separated: https://github.com/home-assistant/home-assistant/pull/4481

danielperna84 commented:
Suggestion for a fallback if lookup via macvendors.com fails (using the [netaddr ](https://pypi.python.org/pypi/netaddr) module, which works offline):
```python
import netaddr
netaddr.EUI("00:00:00:00:00:00").oui.registration().org
```

I'm not aware of the quality of the results the module returns. The bundled [vendor-database](https://github.com/drkjam/netaddr/tree/rel-0.7.x/netaddr/eui) roughly is about a year old.

aequitas commented:
@danielperna84 using netaddr was my first implementation. But @balloob preferred the API. https://github.com/home-assistant/home-assistant/pull/4401

danielperna84 commented:
Oh, ok. Then forget what I've said. :)

aequitas committed:
Add test case for error conditions.

pvizeli commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -460,6 +463,46 @@ def async_update(self):
+                # set something other then None to prevent retry
+                return 'unknown'
+            # request failed, set None to try another time
+            else:
```
Remove that and set except to pass and return None

aequitas replied:
I refactored this bit because the logic didn't make sense. Feel free te review and comment again if needed.

aequitas committed:
There is no reason to retry failes vendor loopups as they won't be saved to the file anyways at that point.

aequitas committed:
Sorry, bad assumption, this only made things worse.

aequitas committed:
Wait for async parts during setup component to complete before asserting results.

aequitas committed:
Fix linting.

aequitas committed:
Is generated when running 'coverage html'.

aequitas commented:
I added changes to fix the ddwrt test suite because the device tracker component setup now has true asynchronous behavior which was unexpected by that test suite.

Also updated the error handling logic to write of any error/failure as `unknown` result. Since the vendor value only gets written to the known_devices file on the first try. So retrying doesn't make sense.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -5,34 +5,37 @@
 
+import aiohttp
+
+import homeassistant.helpers.config_validation as cv
```
Please don't sort imports or clean up code not related to your PR. It makes this PR very hard to code review.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -460,6 +465,46 @@ def async_update(self):
 
+        # lookup vendor string for mac address using api.macvendors.com
+        if not self.vendor:
+            self.vendor = yield from self.get_vendor_for_mac()
```
We only need to look up the vendor when we're writing it to a config file. All other times there is no need to hit this API.

aequitas replied:
If the initial call (before writing to config) fails (not found, error, etc) it will return the value 'unknown' as vendor string. This will be written to config and it won't be retried again. There are test cases for this in the test suite.

aequitas replied:
True, didn't think of that. I will make sure that will be fixed.

balloob commented on tests/test_util/aiohttp.py:
```diff
@@ -64,6 +67,9 @@ def match_request(self, method, url, *, auth=None): \
                 self.mock_calls.append((method, url))
+
+                if self.exc:
+                    raise asyncio.TimeoutError()
```
It would make more sense to raise whatever exception is set in `self.exc`

aequitas replied:
agree, i will refactor.

aequitas committed:
Undo isort.

aequitas committed:
Make aioclient_mock exception more generic.

balloob commented on tests/components/device_tracker/test_init.py:
```diff
@@ -1,27 +1,30 @@
-import os
+from unittest.mock import call, patch
 
-from homeassistant.core import callback
```
Please revert all the changes to the import back so I can see what imports have changed related to this code change.

aequitas replied:
sorry, will do

aequitas committed:
Only lookup mac vendor string with adding new device to known_devices.yaml.

aequitas committed:
Undo isort.

aequitas committed:
Revert unneeded change.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -460,6 +467,47 @@ def async_update(self):
+        url = 'http://api.macvendors.com/' + oui
+        try:
+            with async_timeout.timeout(5, loop=self.hass.loop):
+                resp = yield from self.hass.websession.get(url)
```
We changed the place we store our websession in #4602.

Please use this:

```python
from homeassistant.helpers.aiohttp_client import async_get_clientsession

websession = async_get_clientsession(self.hass)

resp = yield from websession.get(url)
```

balloob reviewed (commented):
Good to go when comment addressed.

aequitas committed:
Adjust to use new websession pattern.

aequitas commented:
Rebased to latests dev and adapted to new websessions pattern.

balloob commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -460,6 +468,49 @@ def async_update(self):
+            websession = async_get_clientsession(self.hass)
+
+            with async_timeout.timeout(5, loop=self.hass.loop):
+                resp = yield from websession.get(url)
```
Can you add a `finally` to make sure the response gets closed?

aequitas replied:
done.

aequitas committed:
Always make sure to cleanup response.

balloob commented:
Ok to merge when merge conflicts resolved 🐬

aequitas committed:
Use correct function to release response.

aequitas committed:
Fix tests.

balloob commented:
Cool!
